### PR TITLE
tests/e2e: Add support to create cluster with operator

### PR DESCRIPTION
**Pull request by kfirtoledo**
Add options to create Clustrlink using the operator in the e2e tests.

### CODE REVIEW

File: tests/e2e/k8s/util/kind.go [hunk 1]
```diff
@@ -151,6 +154,11 @@ func (c *KindCluster) initializeClients() error {
 		return fmt.Errorf("unable to initialize k8s clientset: %w", err)
 	}
 
+	// Add instance CRD to scheme.
```
Comment (orozery) [inline tests/e2e/k8s/util/kind.go:157]:
```suggestion
	// Add clusterlink CRDs to scheme.
```

Also, my personal reference is to rename an import (i.e. `v1alpha1` to `clusterlink`) only when package names collide. But your call.

File: tests/e2e/k8s/util/kind.go [hunk 2]
```diff
@@ -361,6 +369,14 @@ func (c *KindCluster) CreateFromYAML(yaml, namespace string) error {
 		decoder.MutateNamespace(namespace))
 }
 
+// CreateFromFolder creates k8s objects from a yaml in a folder.
+func (c *KindCluster) CreateFromFolder(folder string) error {
```
Comment (orozery) [inline tests/e2e/k8s/util/kind.go:373]: I think folder is a Windows term. Perhaps rename to Directory?

File: tests/e2e/k8s/util/kind.go [hunk 3]
```diff
@@ -361,6 +369,14 @@ func (c *KindCluster) CreateFromYAML(yaml, namespace string) error {
...
+	pattern := "*"
+	fs := os.DirFS(folder)
+	return decoder.DecodeEachFile(context.TODO(), fs, pattern,
```
Comment (orozery) [inline tests/e2e/k8s/util/kind.go:376]:
```suggestion
	return decoder.DecodeEachFile(context.Background(), fs, pattern,
```

File: tests/e2e/k8s/util/kind.go [hunk 4]
```diff
@@ -361,6 +369,14 @@ func (c *KindCluster) CreateFromYAML(yaml, namespace string) error {
 		decoder.MutateNamespace(namespace))
 }
 
+// CreateFromFolder creates k8s objects from a yaml in a folder.
+func (c *KindCluster) CreateFromFolder(folder string) error {
+	pattern := "*"
+	fs := os.DirFS(folder)
```
Comment (orozery) [inline tests/e2e/k8s/util/kind.go:375]: I would remove this lines, and embed the values directly to the function call below.

File: tests/e2e/k8s/util/kind.go [hunk 5]
```diff
@@ -470,6 +486,19 @@ func (c *KindCluster) ScaleDeployment(name, namespace string, replicas int32) er
...
+		return err
+	}
+	return c.CreateFromFolder("./../../../config/operator/rbac/")
+}
+
+// CreateCRDs creates the CRDs from folder config.
+func (c *KindCluster) CreateCRDs() error {
+	return c.CreateFromFolder("./../../../config/operator/crds/")
+}
+
```
Comment (orozery) [inline tests/e2e/k8s/util/kind.go:501]: I'm trying to keep this file to be general kind-utils, and not clusterlink specific.
I would remove those functions, and inline their code in the caller.

File: tests/e2e/k8s/util/fabric.go [hunk 6]
```diff
@@ -146,23 +153,75 @@ func (f *Fabric) SwitchToNewNamespace(name string, appendName bool) error {
...
+
+// deployUsingOperator deploys ClusterLink using operator.
+func (f *Fabric) deployUsingOperator(target *peer, cfg *PeerConfig) error {
+	// Create ClusterLink instance
+	instance, err := f.generateClusterlinkInstance(target, cfg)
+	if err != nil {
+		return fmt.Errorf("cannot generate ClusterLink instance: %w", err)
+	}
+
+	if err := target.cluster.CreateFromYAML(instance, controller.OperatorNameSpace); err != nil {
```
Comment (orozery) [inline tests/e2e/k8s/util/fabric.go:166]: Just noticed this const. NameSpace -> Namespace

**Review by orozery — CHANGES_REQUESTED** [review]

File: tests/e2e/k8s/util/kind.go (hunk 2 above)
Comment (elevran) [reply tests/e2e/k8s/util/kind.go:373]: Or `path`.
Would it also make sense to validate that the parameter is indeed a directory and not a file or let `DirFS` fail?

**Comment by kfirtoledo** [comment]
@orozery PTAL (After the approval I will sync it).

File: pkg/operator/controller/instance_controller.go [hunk 7]
```diff
@@ -47,7 +47,7 @@ const (
 	DataPlaneName     = "cl-dataplane"
 	GoDataPlaneName   = "cl-go-dataplane"
 	IngressName       = "cl-external-svc"
-	OperatorNameSpace = "clusterlink-operator" // TODO -should be removed
+	OperatorNamespace = "clusterlink-operator"
```
Comment (orozery) [inline pkg/operator/controller/instance_controller.go:50]: Why is the default operator namespace different than the instance namespace?

File: tests/e2e/k8s/suite.go [hunk 8]
```diff
@@ -98,8 +98,20 @@ func (s *TestSuite) SetupSuite() {
 		if err != nil {
 			s.T().Fatal(fmt.Errorf("cannot create iperf3-server service: %w", err))
 		}
-	}
+		// create the project operator.
```
Comment (orozery) [inline tests/e2e/k8s/suite.go:101]:
```suggestion

		// create the project operator.
```

File: tests/e2e/k8s/util/fabric.go [hunk 9]
```diff
@@ -20,11 +20,16 @@ import (
...
+	appsv1 "k8s.io/api/apps/v1"
 	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 
 	"github.com/clusterlink-net/clusterlink/pkg/bootstrap"
 	"github.com/clusterlink-net/clusterlink/pkg/bootstrap/platform"
 	"github.com/clusterlink-net/clusterlink/pkg/client"
+	"github.com/clusterlink-net/clusterlink/pkg/operator/controller"
+	"sigs.k8s.io/e2e-framework/klient/wait"
+	"sigs.k8s.io/e2e-framework/klient/wait/conditions"
```
Comment (orozery) [inline tests/e2e/k8s/util/fabric.go:32]: Move these to the upper import group.

File: tests/e2e/k8s/util/fabric.go [hunk 10]
```diff
@@ -37,6 +42,8 @@ type PeerConfig struct {
 	ControlplanePersistency bool
 	// ExpectLargeDataplaneTraffic hints that a large amount of dataplane traffic is expected.
 	ExpectLargeDataplaneTraffic bool
+	// DeployWithOperator deploy cluster using the operator.
```
Comment (orozery) [inline tests/e2e/k8s/util/fabric.go:45]:
```suggestion
	// DeployWithOperator deploys clusterlink using an operator.
```

File: tests/e2e/k8s/util/kind.go [hunk 11]
```diff
@@ -17,11 +17,13 @@ import (
 	"context"
 	"fmt"
 	"io"
+	"os"
 	"reflect"
 	"strings"
 	"sync"
 	"time"
 
+	clusterlink "github.com/clusterlink-net/clusterlink/pkg/apis/clusterlink.net/v1alpha1"
```
Comment (orozery) [inline tests/e2e/k8s/util/kind.go:26]: move this import to its own import group downwards.

**Review by orozery — CHANGES_REQUESTED** [review]

File: pkg/operator/controller/instance_controller.go (hunk 7 above)
Comment (kfirtoledo) [reply pkg/operator/controller/instance_controller.go:50]: Usually, the operator runs in a different namespace with more privileges (like in openshift there is an openshift-operator namespace)

**Review by orozery — APPROVED** [review]

**Pull request merged**